### DISCUSSION TITLE
Sonoff Pow R2 persistent Energy Total when off during midnight

### DISCUSSION THREAD

**Discussion by mikhail5555** [arendst/Tasmota#14283] [Q&A] [answered]
Hey,

From the issue #3689 and the disccussion i started earlier #13781 I understood that its expected that a tasmota device has to stay on 24/7 and is expected to always be on.
In my scenario, my Sonoff Pow R2 lives behind a different switch which is nearly always off.

Due to this, it's not on at midnight, leading to the following code never triggering:
https://github.com/arendst/Tasmota/blob/8fabda4e29d2cc5bbb6bdc390f84d7d62c8b623f/tasmota/xdrv_03_energy.ino#L295-L309

This means, the next day that I turn on my Sonoff Pow R2, the Energy Today gets reset, the Energy Yesterday never gets updated, and the Energy Total doesn't get updated with the energy usage of yesteday, leading to a never correct state.

Is there a way, by a rule or different way, to make the Energy Total & Energy Yesterday being set correctly even though the device is off during midnight?

Might it be an idea to run that piece of code when the device turns on, and there is still a Energy Today (From yesteday) being set in flash (By using the SaveData command for example)?

**Comment by sfromis** [comment, marked as answer]
Well, energy monitoring totals are just useless if you switch off input power to the device, as Tasmota will not keep the saved values updated in flash. Thus you cannot get "correct" values, unless you can restore values from your backend setup. This would be doable, as the message published each TelePeriod has values.

**Reply by mikhail5555** [reply]
Since it's a clean shutdown, I set up a rule when the switch turns to off, it manually calls 'SaveData' to save the values to flash. Thus theoretically it should have access to the correct values before it resets it back to zero due the start of a new day.
